test/extensions/filters/network/ssh/service_userauth_test.cc

```diff
@@ -0,0 +1,746 @@
+  wire::UserAuthInfoResponseMsg resp {
+    .responses = string_list{"response-one", "response-two"},
+  };
+
+  pomerium::extensions::ssh::ClientMessage client_msg;
+  EXPECT_CALL(*transport_, sendMgmtClientMessage(_))
+    .WillOnce(SaveArg<0>(&client_msg));
+
+  auto r = service_->handleMessage(wire::Message{resp});
+  ASSERT_OK(r);
+
+  ASSERT_TRUE(client_msg.has_info_response());
+  auto info_response = client_msg.info_response();
+  pomerium::extensions::ssh::KeyboardInteractiveInfoPromptResponses responses;
+  ASSERT_TRUE(info_response.response().UnpackTo(&responses));
+  ASSERT_EQ(2, responses.responses_size());
+  ASSERT_EQ("response-one", responses.responses()[0]);
+  ASSERT_EQ("response-two", responses.responses()[1]);
+}
+
+TEST_F(DownstreamUserAuthServiceTest, HandleMessageInvalidUserAuthInfoResponse) {
+  // Generate a message with the UserAuthInfoResponse tag but no data.
+  Buffer::OwnedImpl buf;
+  buf.writeBEInt(wire::SshMessageType::UserAuthInfoResponse);
+  wire::Message msg;
+  ASSERT_OK(msg.decode(buf, buf.length()).status());
+
+  auto r = service_->handleMessage(std::move(msg));
+  ASSERT_EQ(r.code(), StatusCode::kInvalidArgument);
+  ASSERT_EQ(r.message(), "invalid auth response");
+}
+
+TEST_F(DownstreamUserAuthServiceTest, HandleMessageSshUnknownType) {
+  auto r = service_->handleMessage(wire::Message{wire::DebugMsg{}});
+  ASSERT_EQ(r.code(), StatusCode::kInternal);
+  ASSERT_EQ(r.message(), "received unexpected message of type Debug (4)");
+}
+
+TEST_F(DownstreamUserAuthServiceTest, HandleMessageSshUnsupportedAuthRequest) {
+  // Construct a UserAuthRequestMsg with an unknown request type. Due to the design of
+  // the sub_message struct this requires decoding from the wire format.
+  Buffer::OwnedImpl buf;
+  static wire::field<std::string, wire::LengthPrefixed> username = "foo@bar"s;
+  static wire::field<std::string, wire::LengthPrefixed> service_name = "ssh-connection"s;
+  static wire::field<std::string, wire::LengthPrefixed> method_name = "UNKNOWN"s;
+  static wire::field<std::string, wire::LengthPrefixed> data = "UNKNOWN-DATA"s;
+  ASSERT_OK(wire::encodeMsg(buf, wire::SshMessageType::UserAuthRequest,
+                                 username,
+                                 service_name,
+                                 method_name,
+                                 data).status());
+  wire::UserAuthRequestMsg user_auth;
+  ASSERT_OK(user_auth.decode(buf, buf.length()).status());
+
+  auto r = service_->handleMessage(wire::Message{user_auth});
+  ASSERT_EQ(r.code(), absl::StatusCode::kUnimplemented);
+  ASSERT_EQ(r.message(), "unknown or unsupported auth method");
+}
+
+TEST_F(DownstreamUserAuthServiceTest, HandleMessageServerAllowUpstream) {
+  auto msg = std::make_unique<pomerium::extensions::ssh::ServerMessage>();
+  auto allow = msg->mutable_auth_response()->mutable_allow();
+  auto upstream = allow->mutable_upstream();
```

Comment:
```suggestion
  auto* allow = msg->mutable_auth_response()->mutable_allow();
  auto* upstream = allow->mutable_upstream();
```

I'd recommend doing this everywhere raw pointers are used with grpc like this